File: day_5/puzzle_2/solve.ts
import fs from "fs";
import { benchmark } from "../../benchmark/benchmark";

benchmark(() => {
  const file = fs.readFileSync("./day_5/input.txt", { encoding: "utf-8" });

  const [stacks, instructions] = file.split("\n\n");

  const stackGrid = stacks
    .split("\n")
    .reverse()
    .reduce(
      (grid, row, i) => {
        if (i === 0) {
          return grid;
        }

        row.split("").forEach((char, j) => {
          if (char === " ") {
            return;
          }

          if ((j - 1) % 4 !== 0) {
            return;
          }

          const position = (j - 1) / 4;

          grid[position].push(char);
        });

        return grid;
      },
      [[], [], [], [], [], [], [], [], []] as string[][],
    );

  instructions
    .replaceAll(/\b(move |from |to )\b/g, "")
    .split("\n")
    .forEach((instruction) => {
      const [move, from, to] = instruction.split(" ").map(Number);

      const crates = stackGrid[from - 1].splice(
        stackGrid[from - 1].length - move,
        move,
      );

      stackGrid[to - 1] = stackGrid[to - 1].concat(crates);
    });

  const topCrates = stackGrid.reduce(
    (string, col) => `${string}${col.at(-1)}`,
    "",
  );

  return topCrates;
}, "day_5_puzzle_2");

File: day_5/puzzle_2/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run do
  file_path = File.join(__dir__, 'input.txt')

  File.open(file_path, 'r') do |file|
    stacks, instructions = file.read.split("\n\n")

    stack_lines = stacks.split("\n").reverse

    stack_grid = stack_lines.each_with_index.each_with_object(Array.new(9) { [] }) do |(row, i), grid|
      next if i.zero?

      row.split('').each_with_index do |char, j|
        next if char == ' '
        next if (j - 1) % 4 != 0

        position = (j - 1) / 4

        grid[position].push(char)
      end
    end

    instructions.gsub(/\b(move |from |to )\b/, '').split("\n").each do |instruction|
      move, from, to = instruction.split(' ').map(&:to_i)

      crates = stack_grid[from - 1].pop(move)

      stack_grid[to - 1].concat(crates)
    end

    top_crates = stack_grid.inject('') { |string, col| "#{string}#{col.last}" }

    top_crates
  end
end
